src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -57,12 +62,128 @@ trait WorkspaceComponent {
     def workspace = foreignKey("FK_WS_ACCESS_WORKSPACE", workspaceId, workspaceQuery)(_.id)
     def group = foreignKey("FK_WS_ACCESS_GROUP", groupName, rawlsGroupQuery)(_.groupName)
 
-    def x = primaryKey("PK_WORKSPACE_ACCESS", (workspaceId, accessLevel))
+    def accessPrimaryKey = primaryKey("PK_WORKSPACE_ACCESS", (workspaceId, accessLevel))
 
     def * = (workspaceId, groupName, accessLevel) <> (WorkspaceAccessRecord.tupled, WorkspaceAccessRecord.unapply)
   }
 
-  protected val workspaceQuery = TableQuery[WorkspaceTable]
   protected val workspaceAttributeQuery = TableQuery[WorkspaceAttributeTable]
   protected val workspaceAccessQuery = TableQuery[WorkspaceAccessTable]
+
+  object workspaceQuery extends TableQuery(new WorkspaceTable(_)) {
+    def workspaceAccessGroups(workspaceId: UUID) = for {
+      workspaceRec <- workspaceQuery if workspaceRec.id === workspaceId
+      workspaceAccessRec <- workspaceAccessQuery if workspaceRec.id === workspaceAccessRec.workspaceId
+    } yield (workspaceAccessRec)
+
+    def save(workspace: Workspace): DBIOAction[Workspace, NoStream, Effect.Read with Effect.Write] = {
+      val workspaceRecord = marshalWorkspace(workspace)
+
+      workspaceQuery insertOrUpdate workspaceRecord andThen {
+        DBIO.seq(workspace.accessLevels.map { case (accessLevel, group) =>
+          workspaceAccessQuery insertOrUpdate WorkspaceAccessRecord(workspaceRecord.id, group.groupName.value, accessLevel.toString)
+        }.toSeq: _*)
+
+      } andThen {
+        workspaceAttributes(workspaceRecord.id).result.flatMap { attributeRecords =>
+          // clear existing attributes, any concurrency locking should be handled at the workspace level
+          val deleteActions = deleteWorkspaceAttributes(attributeRecords)
+
+          // insert attributes
+          val insertActions = workspace.attributes.flatMap { case (name, attribute) =>
+            attributeQuery.insertAttributeRecords(name, attribute, workspaceRecord.id).map(_.flatMap{ attrId =>
+              insertWorkspaceAttributeMapping(workspaceRecord, attrId)
+            })
+          }
+
+          DBIO.seq(deleteActions ++ insertActions:_*)
+        }
+      } map(_ => workspace)
+    }
+
+    def findByName(workspaceName: WorkspaceName): DBIOAction[Option[Workspace], NoStream, Effect.Read] = {
+      loadWorkspace(findByNameQuery(workspaceName))
+    }
+
+    def findById(workspaceId: String): DBIOAction[Option[Workspace], NoStream, Effect.Read] = {
+      loadWorkspace(findByIdQuery(UUID.fromString(workspaceId)))
+    }
+
+    def delete(workspaceName: WorkspaceName): DBIOAction[Boolean, NoStream, Effect.Read with Effect.Write] = {
+      findByNameQuery(workspaceName).result.flatMap {
+        case Seq() => DBIO.successful(false)
+        case Seq(workspaceRecord) =>
+          workspaceAttributes(workspaceRecord.id).result.flatMap(recs => DBIO.seq(deleteWorkspaceAttributes(recs):_*)) flatMap { _ =>
```

Comment:
andThen is merely ordering where the second thing does not care about the results of the first thing. flatMap (and map) also does the ordering but passes the result of the first part to the second part.